Persist ExposureWindow based risk level calculation results and update the UI (EXPOSUREAPP-3910,EXPOSUREAPP-3855)

d4rken opened the pull request:
Combined efforts of @chris-cwa and me to rid ourselves of the `RisklevelRepository` and make UI updates more consistent by providing a single-source-of-truth for risk level related data.

* Remove `RiskLevelRespository`
* Replace static access to `LocalData.*risklevel*`
* Refactor all access to the results produced by the  `RiskLevelTask`. The task stores the data in `RiskLevelStorage`, and any component wants the calculated score can get the latest scores via `RiskLevelStorage.latestResults: Flow<List<RiskLevelResult>>`.
    * UI Components subscribe to changes in `RiskLevelStorage`
    * Notifications are displayed via `RiskLevelChangeDetector` by monitoring new `RiskLevelResult`s

### TODO:
* [x] Tests
* [x] Switch to new UI texts added bei #1697 
* [x] Legacy data migration
* [x] Persist ExposureWindow's in tester builds.


### Testing
* Use the debug menu to fake a high/low/default risklevel state and  check that the riskcard looks like the screenshots in EXPOSUREAPP-3855
* Test general risk level behavior on fresh installs and when time traveling


New risk card examples:

![Screenshot from 2020-11-24 12-58-40](https://user-images.githubusercontent.com/1439229/100091492-f6f03500-2e54-11eb-94e8-055b5be82ec2.png)
![Screenshot from 2020-11-24 12-58-23](https://user-images.githubusercontent.com/1439229/100091498-f788cb80-2e54-11eb-8c46-e49ecd5b8125.png)
![Screenshot from 2020-11-24 12-58-10](https://user-images.githubusercontent.com/1439229/100091499-f788cb80-2e54-11eb-8de3-44dad3527791.png)

d4rken committed:
First draft to refactor `RiskRepository` away and persist risk level results produced by ExposureWindow calculations.

TODO: Tests, Legacy data migration, Persist ExposureWindow's in tester builds.

d4rken committed:
Unit tests for refactored classes.

d4rken committed:
LINTs

d4rken committed:
First draft for legacy risk data migration.

TODO: Tests.

chris-cwa commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/storage/legacy/RiskLevelResultMigrator.kt:
```diff
@@ -1,137 +1,64 @@
-     ****************************************************/
+    private fun lastTimeRiskLevelCalculation(): Instant? {
+        val time = prefs.getLong("preference_timestamp_risk_level_calculation", 0L)
+        return Instant.ofEpochMilli(time)
```
if (time == 0) return null

d4rken committed:
Store exposure windows on device for tester builds and perform clean on them too after risk result deletions.
Split RiskResultStorage such that the exposure window storage code is only available in deviceForTesters flavor builds.

chris-cwa committed:
simplification

chris-cwa committed:
no preference should return no value

d4rken committed:
Merge branch 'release/1.8.x' into feature/3855-exposure-windows-UI-changes

d4rken committed:
Remove TODOs, singletons are refactored away ;)

d4rken committed:
Simplify RiskLevelTask interface

d4rken committed:
Split risk level storage implementations to let production (device flavor) have a NOOP for storing/deleting exposure windows.

d4rken committed:
Remove unused storage function.
Add test skeletons.

chris-cwa committed:
unit test

chris-cwa committed:
unit test

d4rken committed:
Merge branch 'release/1.8.x' into feature/3855-exposure-windows-UI-changes

# Conflicts:
#	Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt

d4rken committed:
Merge remote-tracking branch 'origin/feature/3855-exposure-windows-UI-changes' into feature/3855-exposure-windows-UI-changes

chris-cwa committed:
Merge branch 'feature/3855-exposure-windows-UI-changes' of https://github.com/corona-warn-app/cwa-app-android into feature/3855-exposure-windows-UI-changes

d4rken committed:
Update risk card to show new window mode based information.

d4rken committed:
Merge remote-tracking branch 'origin/feature/3855-exposure-windows-UI-changes' into feature/3855-exposure-windows-UI-changes

d4rken committed:
Fix test regressions.

kolyaopahle commented on Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt:
```diff
@@ -129,26 +135,24 @@ class TestRiskLevelCalculationFragmentCWAViewModel @AssistedInject constructor(
-        exposureResultStore.matchedKeyCount,
-        exposureResultStore.daysSinceLastExposure,
+        riskLevelStorage.riskLevelResults,
         LocalData.lastTimeDiagnosisKeysFromServerFetchFlow()
```
This also feels like something that could be part of riskLevelStorage?

d4rken replied:
Hm, it's not risk level related though. Downloading and calculation is independent.
In any case, the actual timestamps are already tracked by `KeyPackageSyncSettings`, we could possibly remove the `LocalData.lastTimeDiagnosisKeys` soon, let's evaluate it after all PRs are merged for 1.8.x, what do you think?

kolyaopahle replied:
hmm true, maybe riskLevelStorage is the wrong place, reevaluation after the 1.8 merge sounds good to me :)

kolyaopahle commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/ExposureStateUpdateWorker.kt:
```diff
@@ -19,18 +18,18 @@ import timber.log.Timber
-                exposureResultStore.entities.value = ExposureResult(it, null)
-                Timber.v("exposure summary state updated: $it")
-            }
+//            enfClient.exposureWindows().let {
```
Why is this commented out and not removed?

d4rken replied:
Oversight. Fixed in next commit :+1:

kolyaopahle commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelChangeDetector.kt:
```diff
@@ -0,0 +1,102 @@
+) {
+
+    fun launch() {
+        Timber.v("Monitoring config changes.")
```
Is this actually monitoring the app config?

d4rken replied:
Copy paste ;)

kolyaopahle commented on Corona-Warn-App/src/test/java/de/rki/coronawarnapp/risk/RiskLevelDataTest.kt:
```diff
@@ -22,7 +22,7 @@ class RiskLevelDataTest : BaseTest() {
     }
 
-    fun createInstance() = RiskLevelData(context = context)
+    fun createInstance() = RiskLevelSettings(context = context)
```
since we are actually testing RiskLevelSettings maybe rename/refactor the test?

harambasicluka commented on Corona-Warn-App/src/main/res/values-bg/strings.xml:
```diff
@@ -34,12 +34,6 @@
     <!-- NOTR -->
     <string name="preference_initial_result_received_time"><xliff:g id="preference">"preference_initial_result_received_time"</xliff:g></string>
     <!-- NOTR -->
-    <string name="preference_risk_level_score"><xliff:g id="preference">"preference_risk_level_score"</xliff:g></string>
```
shouldn't be excluded to not cause any merge conflicts as the translation delivery will "override" everything. In the newest translation delivery after this is merged this shouldn't be included. But as we'll get our next translation delivery tomorrow I would say we leave it as it is.

harambasicluka replied:
We may have to do resolve some merge conflicts on the translation delivery.

harambasicluka replied:
This applies for all languages != de

harambasicluka reviewed (commented):
Added some questions, have to review the tests.

Of course naming is highly optional, but I would prefer if we could stick to `HIGH` or `INCREASED` Risk in our code.

harambasicluka reviewed (changes requested):
Updated 1.7.0/RC8 to this branch, App crashed instantly.

```
E/AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-4 @coroutine#69
    Process: de.rki.coronawarnapp.test, PID: 31644
    java.lang.NoClassDefFoundError: Failed resolution of: Lde/rki/coronawarnapp/risk/storage/legacy/RiskLevelResultMigrator$getLegacyResults$1$1;
        at de.rki.coronawarnapp.risk.storage.legacy.RiskLevelResultMigrator.getLegacyResults(RiskLevelResultMigrator.kt:48)
        at de.rki.coronawarnapp.risk.storage.BaseRiskLevelStorage$$special$$inlined$map$3$2.emit(Collect.kt:136)
        at de.rki.coronawarnapp.risk.storage.BaseRiskLevelStorage$$special$$inlined$map$2$2.emit(Collect.kt:136)
        at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:15)
        at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(Unknown Source:4)
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:73)
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:55)
        at androidx.room.CoroutinesRoom$Companion$createFlow$1$1$1.invokeSuspend(CoroutinesRoom.kt:82)
        (Coroutine boundary)
        at androidx.room.CoroutinesRoom$Companion$createFlow$1$1.invokeSuspend(CoroutinesRoom.kt:82)
        at androidx.room.CoroutinesRoom$Companion$createFlow$1.invokeSuspend(CoroutinesRoom.kt:75)
        at kotlinx.coroutines.flow.AbstractFlow.collect(Flow.kt:212)
        at kotlinx.coroutines.flow.internal.CombineKt$asFairChannel$1.invokeSuspend(Combine.kt:143)
        at kotlinx.coroutines.flow.FlowKt__EmittersKt$onStart$$inlined$unsafeFlow$1.collect(Emitters.kt:120)
        at kotlinx.coroutines.flow.FlowKt__EmittersKt$onCompletion$$inlined$unsafeFlow$1.collect(Emitters.kt:114)
        at androidx.lifecycle.FlowLiveDataConversions$asLiveData$1.invokeSuspend(FlowLiveData.kt:139)
        at androidx.lifecycle.BlockRunner$maybeRun$1.invokeSuspend(CoroutineLiveData.kt:176)
     Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Lde/rki/coronawarnapp/risk/storage/legacy/RiskLevelResultMigrator$getLegacyResults$1$1;
        at de.rki.coronawarnapp.risk.storage.legacy.RiskLevelResultMigrator.getLegacyResults(RiskLevelResultMigrator.kt:48)
        at de.rki.coronawarnapp.risk.storage.BaseRiskLevelStorage$$special$$inlined$map$3$2.emit(Collect.kt:136)
        at de.rki.coronawarnapp.risk.storage.BaseRiskLevelStorage$$special$$inlined$map$2$2.emit(Collect.kt:136)
        at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:15)
        at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(Unknown Source:4)
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:73)
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:55)
        at androidx.room.CoroutinesRoom$Companion$createFlow$1$1$1.invokeSuspend(CoroutinesRoom.kt:82)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "de.rki.coronawarnapp.risk.storage.legacy.RiskLevelResultMigrator$getLegacyResults$1$1" on path: DexPathList[[zip file "/data/app/~~fvwqdjrF-VLWwJljB4CPig==/de.rki.coronawarnapp.test-_ENM36BVz0C2KLhOLp3G2w==/base.apk"],nativeLibraryDirectories=[/data/app/~~fvwqdjrF-VLWwJljB4CPig==/de.rki.coronawarnapp.test-_ENM36BVz0C2KLhOLp3G2w==/lib/arm64, /data/app/~~fvwqdjrF-VLWwJljB4CPig==/de.rki.coronawarnapp.test-_ENM36BVz0C2KLhOLp3G2w==/base.apk!/lib/arm64-v8a, /system/lib64, /system_ext/lib64, /product/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:207)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at de.rki.coronawarnapp.risk.storage.legacy.RiskLevelResultMigrator.getLegacyResults(RiskLevelResultMigrator.kt:48) 
        at de.rki.coronawarnapp.risk.storage.BaseRiskLevelStorage$$special$$inlined$map$3$2.emit(Collect.kt:136) 
        at de.rki.coronawarnapp.risk.storage.BaseRiskLevelStorage$$special$$inlined$map$2$2.emit(Collect.kt:136) 
        at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:15) 
        at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(Unknown Source:4) 
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:73) 
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:55) 
        at androidx.room.CoroutinesRoom$Companion$createFlow$1$1$1.invokeSuspend(CoroutinesRoom.kt:82) 
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665) 
```

harambasicluka commented:
App is now opening, but the the contact row is missing:
![image](https://user-images.githubusercontent.com/64483219/100111284-3aa36880-2e6e-11eb-9723-96c49bf9f849.png)

harambasicluka commented:
> App is now opening, but the the contact row is missing:

Still get the same state after a clean install with time traveling to get a low risk.

harambasicluka commented:
> > App is now opening, but the the contact row is missing:
> 
> Still get the same state after a clean install with time traveling to get a low risk.

Is "fixed" english strings are empty... you have to use german as your device language :D

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelChangeDetector.kt:
```diff
@@ -0,0 +1,102 @@
+            Timber.d("Notification Permission = ${notificationManagerCompat.areNotificationsEnabled()}")
+
+            if (!foregroundState.isInForeground.first()) {
+                NotificationHelper.sendNotification("", context.getString(R.string.notification_body), true)
```
Was previously like that, didn't touch that part.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelChangeDetector.kt:
```diff
@@ -0,0 +1,102 @@
+            previous.isHighRisk != current.isHighRisk
+
+        private val HIGH_RISK_LEVELS = arrayOf(RiskLevel.INCREASED_RISK)
+        private val LOW_RISK_LEVELS = arrayOf( // TODO delete?
```
Yes

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt:
```diff
@@ -40,66 +34,86 @@ import timber.log.Timber
-                    } else {
-                        UNKNOWN_RISK_OUTDATED_RESULTS_MANUAL
-                    }
+            return kotlin.run evaluation@{
```
It's a scope label.

https://kotlinlang.org/docs/reference/this-expressions.html

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt:
```diff
@@ -40,66 +34,86 @@ import timber.log.Timber
+            return kotlin.run evaluation@{
+                if (calculationNotPossibleBecauseOfNoKeys()) {
+                    return@evaluation RiskLevelTaskResult(
+                        riskLevel = UNKNOWN_RISK_INITIAL,
```
Will sync with @SamuraiKek and solve the conflicts, let's merge this first.

d4rken committed:
Address PR comments.

d4rken committed:
LINTs

harambasicluka commented:
Discussed with @d4rken my questions and bugs are resolved. As I won't be available this evening you can see this PR as approved from my side when Circle Ci passes.

d4rken committed:
Everybody get's LINT for XMAS.

d4rken committed:
Finish unit tests and remove unused classes.

kolyaopahle reviewed (approved):
LGTM 👍

BMItter commented on Corona-Warn-App/src/test/java/de/rki/coronawarnapp/risk/RiskLevelResultTest.kt:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @Test
+    fun testUnsuccessfulRistLevels() {
```
little typo here- ristLevel 
but no urgent

BMItter commented:
Did a deeplook yesterday evening. Code lgtm & could not find any bugs :ok_hand: . left one little comment (optional).
